File: python/src/debiased_inference/sets.py
"""Grid-based density level sets and inverse-regression sets."""

from __future__ import annotations

from statistics import NormalDist

import numpy as np
from numpy.typing import ArrayLike

from ._results import ConfidenceBandResult, EstimateResult, SetEstimateResult
from ._validation import (
    as_evaluation_points,
    as_samples,
    as_vector,
    bootstrap_parameters,
    positive_scalar,
)


def _crossing_roots(points: np.ndarray, values: np.ndarray, level: float) -> np.ndarray:
    order = np.argsort(points)
    x = points[order]
    shifted = values[order] - level
    roots: list[float] = []
    exact = np.flatnonzero(shifted == 0)
    roots.extend(float(x[index]) for index in exact)
    crossing = np.flatnonzero(shifted[:-1] * shifted[1:] < 0)
    for index in crossing:
        fraction = -shifted[index] / (shifted[index + 1] - shifted[index])
        roots.append(float(x[index] + fraction * (x[index + 1] - x[index])))
    return np.unique(np.asarray(roots, dtype=float))


def _contour_points(
    points: np.ndarray, values: np.ndarray, level: float
) -> np.ndarray:
    """Approximate a 2-D contour on a complete rectangular grid."""
    x_coordinates = np.unique(points[:, 0])
    y_coordinates = np.unique(points[:, 1])
    if points.shape[0] != x_coordinates.size * y_coordinates.size:
        raise ValueError("two-dimensional level sets require a complete rectangular grid")
    surface = np.full((x_coordinates.size, y_coordinates.size), np.nan)
    x_index = np.searchsorted(x_coordinates, points[:, 0])
    y_index = np.searchsorted(y_coordinates, points[:, 1])
    surface[x_index, y_index] = values
    if np.any(~np.isfinite(surface)):
        raise ValueError("two-dimensional level sets require unique rectangular grid points")

    intersections: list[np.ndarray] = []
    for i in range(x_coordinates.size - 1):
        for j in range(y_coordinates.size - 1):
            vertices = np.array(
                [
                    [x_coordinates[i], y_coordinates[j]],
                    [x_coordinates[i + 1], y_coordinates[j]],
                    [x_coordinates[i + 1], y_coordinates[j + 1]],
                    [x_coordinates[i], y_coordinates[j + 1]],
                ]
            )
            heights = np.array(
                [
                    surface[i, j],
                    surface[i + 1, j],
                    surface[i + 1, j + 1],
                    surface[i, j + 1],
                ]
            )
            for triangle in ((0, 1, 2), (0, 2, 3)):
                triangle_points = vertices[list(triangle)]
                shifted = heights[list(triangle)] - level
                found: list[np.ndarray] = []
                for left, right in ((0, 1), (1, 2), (2, 0)):
                    if shifted[left] == 0:
                        found.append(triangle_points[left])
                    if shifted[left] * shifted[right] < 0:
                        fraction = -shifted[left] / (shifted[right] - shifted[left])
                        found.append(
                            triangle_points[left]
                            + fraction
                            * (triangle_points[right] - triangle_points[left])
                        )
                if shifted[2] == 0:
                    found.append(triangle_points[2])
                if len(found) >= 2:
                    intersections.extend(found[:2])
    if not intersections:
        return np.empty((0, 2), dtype=float)
    return np.unique(np.round(np.vstack(intersections), decimals=14), axis=0)


def _level_geometry(points: np.ndarray, values: np.ndarray, level: float) -> np.ndarray:
    if points.ndim == 1:
        return _crossing_roots(points, values, level)
    if points.ndim == 2 and points.shape[1] == 2:
        return _contour_points(points, values, level)
    raise ValueError("level-set geometry is supported for one or two dimensions")


def _point_cloud(values: ArrayLike, *, name: str) -> np.ndarray:
    result = np.asarray(values, dtype=float)
    if result.ndim == 1:
        result = result[:, None]
    if result.ndim != 2 or result.shape[0] == 0 or result.shape[1] == 0:
        raise ValueError(f"{name} must be a non-empty vector or point matrix")
    if not np.all(np.isfinite(result)):
        raise ValueError(f"{name} must contain only finite values")
    return result


def _distance_to_geometry(points: np.ndarray, geometry: np.ndarray) -> np.ndarray:
    point_cloud = points[:, None] if points.ndim == 1 else points
    geometry_cloud = geometry[:, None] if geometry.ndim == 1 else geometry
    differences = point_cloud[:, None, :] - geometry_cloud[None, :, :]
    return np.min(np.sqrt(np.sum(differences**2, axis=-1)), axis=1)


def level_set(estimate: EstimateResult, level: float) -> SetEstimateResult:
    """Estimate a 1-D or regular-grid 2-D equality level set."""
    points = np.asarray(estimate.points, dtype=float)
    if (
        points.ndim not in {1, 2}
        or points.shape[0] == 0
        or not np.all(np.isfinite(points))
    ):
        raise ValueError("estimate.points must be a finite vector or point matrix")
    level = float(level)
    if not np.isfinite(level):
        raise ValueError("level must be finite")
    roots = _level_geometry(points, estimate.estimate, level)
    return SetEstimateResult(
        points=points.copy(),
        mask=np.zeros(points.size, dtype=bool),
        roots=roots,
        level=level,
        method="level_set",
    )


def invert_confidence_band(
    band: ConfidenceBandResult, level: float
) -> SetEstimateResult:
    """Invert a simultaneous band to obtain a confidence set for a level set.

    This implements the alternative construction in Remark 2 for density and
    its regression analogue in Section 3.2.1.
    """
    points = np.asarray(band.points, dtype=float)
    if (
        points.ndim not in {1, 2}
        or points.shape[0] == 0
        or not np.all(np.isfinite(points))
    ):
        raise ValueError("band.points must be a finite vector or point matrix")
    level = float(level)
    if not np.isfinite(level):
        raise ValueError("level must be finite")
    mask = (band.lower <= level) & (level <= band.upper)
    roots = _level_geometry(points, band.estimate, level)
    return SetEstimateResult(
        points=points.copy(),
        mask=mask,
        roots=roots,
        level=level,
        method="band_inversion",
        confidence=band.confidence,
        n_boot=band.n_boot,
        bootstrap_statistics=band.bootstrap_statistics.copy(),
    )


def hausdorff_distance(a: ArrayLike, b: ArrayLike) -> float:
    """Hausdorff distance between finite point clouds of equal dimension."""
    left = _point_cloud(a, name="a")
    right = _point_cloud(b, name="b")
    if left.shape[1] != right.shape[1]:
        raise ValueError("a and b must have the same point dimension")
    differences = left[:, None, :] - right[None, :, :]
    pairwise = np.sqrt(np.sum(differences**2, axis=-1))
    return float(max(np.max(np.min(pairwise, axis=1)), np.max(np.min(pairwise, axis=0))))


def density_level_set(
    x: ArrayLike,
    level: float,
    points: ArrayLike | None = None,
    *,
    bandwidth: float | None = None,
    tau: float = 1.0,
    grid_size: int = 200,
) -> SetEstimateResult:
    """Estimate a 1-D or regular-grid 2-D density equality level set."""
    from .kde import debiased_kde

    return level_set(
        debiased_kde(
            x, points, bandwidth=bandwidth, tau=tau, grid_size=grid_size
        ),
        level,
    )


def density_level_set_confidence(
    x: ArrayLike,
    level: float,
    points: ArrayLike | None = None,
    *,
    bandwidth: float | None = None,
    tau: float = 1.0,
    confidence: float = 0.95,
    n_boot: int = 999,
    method: str = "hausdorff",
    random_state: int | None = None,
    grid_size: int = 200,
    max_attempts: int | None = None,
) -> SetEstimateResult:
    """Construct a confidence set for a 1-D or regular-grid 2-D density level set.

    ``method="hausdorff"`` implements Section 3.1.1 by bootstrapping the
    Hausdorff distance between interpolated level-set roots. ``"inversion"``
    implements Remark 2 by inverting the simultaneous density band.
    """
    from .bandwidth import density_bandwidth
    from .kde import _estimate_from_matrix, _kernel_matrix, kde_confidence_band

    if method not in {"hausdorff", "inversion"}:
        raise ValueError("method must be 'hausdorff' or 'inversion'")
    if method == "inversion":
        return invert_confidence_band(
            kde_confidence_band(
                x,
                points,
                bandwidth=bandwidth,
                tau=tau,
                confidence=confidence,
                n_boot=n_boot,
                random_state=random_state,
                grid_size=grid_size,
            ),
            level,
        )

    confidence, n_boot = bootstrap_parameters(confidence, n_boot)
    samples = as_samples(x)
    if samples.shape[1] not in {1, 2}:
        raise ValueError("Hausdorff level-set confidence supports one or two dimensions")
    evaluation_matrix = as_evaluation_points(points, samples, grid_size=grid_size)
    evaluation = (
        evaluation_matrix[:, 0]
        if samples.shape[1] == 1
        else evaluation_matrix.copy()
    )
    selected = density_bandwidth(samples) if bandwidth is None else positive_scalar(
        bandwidth, name="bandwidth"
    )
    tau = positive_scalar(tau, name="tau")
    level = float(level)
    if not np.isfinite(level):
        raise ValueError("level must be finite")
    matrix = _kernel_matrix(samples, evaluation_matrix, selected, tau)
    estimate = _estimate_from_matrix(matrix, selected, samples.shape[1])
    roots = _level_geometry(evaluation, estimate, level)
    if roots.size == 0:
        raise ValueError("estimated level set is empty on the evaluation grid")

    generator = np.random.default_rng(random_state)
    attempts_limit = max_attempts if max_attempts is not None else max(10 * n_boot, 100)
    if not isinstance(attempts_limit, (int, np.integer)) or attempts_limit < n_boot:
        raise ValueError("max_attempts must be an integer at least n_boot")
    statistics = np.empty(n_boot, dtype=float)
    accepted = 0
    attempts = 0
    probabilities = np.full(samples.shape[0], 1.0 / samples.shape[0])
    while accepted < n_boot and attempts < attempts_limit:
        attempts += 1
        counts = generator.multinomial(samples.shape[0], probabilities)
        bootstrap_estimate = (
            matrix @ counts
            / (samples.shape[0] * selected ** samples.shape[1])
        )
        bootstrap_roots = _level_geometry(evaluation, bootstrap_estimate, level)
        if bootstrap_roots.size:
            statistics[accepted] = hausdorff_distance(bootstrap_roots, roots)
            accepted += 1
    if accepted < n_boot:
        raise RuntimeError(
            f"only {accepted} non-empty bootstrap level sets after {attempts} attempts"
        )
    radius = float(np.quantile(statistics, confidence))
    mask = _distance_to_geometry(evaluation, roots) <= radius
    return SetEstimateResult(
        points=evaluation.copy(),
        mask=mask,
        roots=roots,
        level=level,
        method="hausdorff",
        radius=radius,
        confidence=confidence,
        n_boot=n_boot,
        bootstrap_statistics=statistics,
    )


def inverse_regression(
    x: ArrayLike,
    y: ArrayLike,
    level: float,
    points: ArrayLike | None = None,
    *,
    bandwidth: float | None = None,
    tau: float = 1.0,
    grid_size: int = 200,
    n_folds: int = 5,
    random_state: int | None = 0,
) -> SetEstimateResult:
    """Estimate a one-dimensional inverse-regression equality set."""
    from .regression import debiased_local_linear

    return level_set(
        debiased_local_linear(
            x,
            y,
            points,
            bandwidth=bandwidth,
            tau=tau,
            grid_size=grid_size,
            n_folds=n_folds,
            random_state=random_state,
        ),
        level,
    )


def inverse_regression_confidence(
    x: ArrayLike,
    y: ArrayLike,
    level: float,
    points: ArrayLike | None = None,
    *,
    bandwidth: float | None = None,
    tau: float = 1.0,
    confidence: float = 0.95,
    n_boot: int = 999,
    method: str = "hausdorff",
    random_state: int | None = None,
    grid_size: int = 200,
    n_folds: int = 5,
    max_attempts: int | None = None,
) -> SetEstimateResult:
    """Construct a confidence set for an inverse-regression equality set."""
    from .regression import (
        _debiased_regression_values,
        debiased_local_linear,
        regression_confidence_band,
    )

    if method not in {"hausdorff", "inversion", "normal"}:
        raise ValueError("method must be 'hausdorff', 'inversion', or 'normal'")
    if method == "inversion":
        return invert_confidence_band(
            regression_confidence_band(
                x,
                y,
                points,
                bandwidth=bandwidth,
                tau=tau,
                confidence=confidence,
                n_boot=n_boot,
                random_state=random_state,
                grid_size=grid_size,
                n_folds=n_folds,
                max_attempts=max_attempts,
            ),
            level,
        )

    confidence, n_boot = bootstrap_parameters(confidence, n_boot)
    x_values = as_vector(x, name="x")
    y_values = as_vector(y, name="y")
    if x_values.size != y_values.size or x_values.size < 4:
        raise ValueError("x and y must have equal lengths of at least 4")
    base = debiased_local_linear(
        x_values,
        y_values,
        points,
        bandwidth=bandwidth,
        tau=tau,
        grid_size=grid_size,
        n_folds=n_folds,
        random_state=random_state,
    )
    level = float(level)
    if not np.isfinite(level):
        raise ValueError("level must be finite")
    roots = _crossing_roots(base.points, base.estimate, level)
    if roots.size == 0:
        raise ValueError("estimated inverse-regression set is empty on the evaluation grid")
    if method == "normal" and roots.size != 1:
        raise ValueError("method='normal' requires exactly one estimated crossing")
    if method == "normal" and n_boot < 2:
        raise ValueError("method='normal' requires at least two bootstrap replicates")

    generator = np.random.default_rng(random_state)
    attempts_limit = max_attempts if max_attempts is not None else max(10 * n_boot, 100)
    if not isinstance(attempts_limit, (int, np.integer)) or attempts_limit < n_boot:
        raise ValueError("max_attempts must be an integer at least n_boot")
    statistics = np.empty(n_boot, dtype=float)
    accepted = 0
    attempts = 0
    while accepted < n_boot and attempts < attempts_limit:
        attempts += 1
        indices = generator.integers(0, x_values.size, size=x_values.size)
        estimate = _debiased_regression_values(
            x_values[indices],
            y_values[indices],
            base.points,
            base.bandwidth,
            base.tau,
        )
        if np.all(np.isfinite(estimate)):
            bootstrap_roots = _crossing_roots(base.points, estimate, level)
            valid_roots = (
                bootstrap_roots.size == 1
                if method == "normal"
                else bootstrap_roots.size > 0
            )
            if valid_roots:
                statistics[accepted] = (
                    bootstrap_roots[0]
                    if method == "normal"
                    else hausdorff_distance(bootstrap_roots, roots)
                )
                accepted += 1
    if accepted < n_boot:
        raise RuntimeError(
            f"only {accepted} non-empty bootstrap inverse sets after {attempts} attempts"
        )
    if method == "normal":
        standard_error = float(np.std(statistics, ddof=1))
        radius = NormalDist().inv_cdf(0.5 + confidence / 2.0) * standard_error
    else:
        radius = float(np.quantile(statistics, confidence))
    mask = np.min(np.abs(base.points[:, None] - roots[None, :]), axis=1) <= radius
    return SetEstimateResult(
        points=base.points.copy(),
        mask=mask,
        roots=roots,
        level=level,
        method=method,
        radius=radius,
        confidence=confidence,
        n_boot=n_boot,
        bootstrap_statistics=statistics,
    )

File: r/R/sets.R
.crossing_roots <- function(points, values, level) {
  ordering <- order(points)
  x <- points[ordering]
  shifted <- values[ordering] - level
  roots <- x[shifted == 0]
  crossing <- which(
    shifted[seq_len(length(shifted) - 1L)] * shifted[-1L] < 0
  )
  for (index in crossing) {
    fraction <- -shifted[index] / (shifted[index + 1L] - shifted[index])
    roots <- c(roots, x[index] + fraction * (x[index + 1L] - x[index]))
  }
  sort(unique(as.double(roots)))
}

.contour_points <- function(points, values, level) {
  x_coordinates <- sort(unique(points[, 1L]))
  y_coordinates <- sort(unique(points[, 2L]))
  if (nrow(points) != length(x_coordinates) * length(y_coordinates)) {
    stop("two-dimensional level sets require a complete rectangular grid",
         call. = FALSE)
  }
  surface <- matrix(NA_real_, nrow = length(x_coordinates),
                    ncol = length(y_coordinates))
  x_index <- match(points[, 1L], x_coordinates)
  y_index <- match(points[, 2L], y_coordinates)
  surface[cbind(x_index, y_index)] <- values
  if (any(!is.finite(surface))) {
    stop("two-dimensional level sets require unique rectangular grid points",
         call. = FALSE)
  }
  lines <- grDevices::contourLines(
    x = x_coordinates, y = y_coordinates, z = surface, levels = level
  )
  if (length(lines) == 0L) return(matrix(numeric(), nrow = 0L, ncol = 2L))
  geometry <- do.call(rbind, lapply(lines, function(line) cbind(line$x, line$y)))
  unique(round(geometry, digits = 14L))
}

.level_geometry <- function(points, values, level) {
  if (is.null(dim(points))) return(.crossing_roots(points, values, level))
  if (is.matrix(points) && ncol(points) == 2L) {
    return(.contour_points(points, values, level))
  }
  stop("level-set geometry is supported for one or two dimensions", call. = FALSE)
}

.as_set_points <- function(x, name) {
  if (is.vector(x) && is.numeric(x)) x <- matrix(x, ncol = 1L)
  if (is.data.frame(x)) x <- as.matrix(x)
  if (!is.matrix(x) || !is.numeric(x) || nrow(x) == 0L || ncol(x) == 0L) {
    stop(name, " must be a non-empty vector or point matrix", call. = FALSE)
  }
  storage.mode(x) <- "double"
  if (any(!is.finite(x))) stop(name, " must contain only finite values", call. = FALSE)
  x
}

.distance_to_geometry <- function(points, geometry) {
  point_cloud <- if (is.null(dim(points))) matrix(points, ncol = 1L) else points
  geometry_cloud <- if (is.null(dim(geometry))) matrix(geometry, ncol = 1L) else geometry
  apply(point_cloud, 1L, function(point) {
    min(sqrt(rowSums(sweep(geometry_cloud, 2L, point, "-")^2)))
  })
}

#' Estimate a one- or two-dimensional equality level set
#' @param estimate An object returned by `debiased_kde()` or
#'   `debiased_local_linear()`.
#' @param level Finite target level.
#' @return An object of class `di_set`.
#' @export
level_set <- function(estimate, level) {
  if (!inherits(estimate, "di_estimate")) {
    stop("estimate must be a di_estimate object", call. = FALSE)
  }
  points <- estimate$points
  if (is.null(dim(points))) {
    points <- .as_vector(points, "estimate$points")
  } else {
    points <- .as_set_points(points, "estimate$points")
  }
  if (!is.numeric(level) || length(level) != 1L || !is.finite(level)) {
    stop("level must be finite", call. = FALSE)
  }
  structure(list(
    points = points, mask = rep(FALSE, length(estimate$estimate)),
    roots = .level_geometry(points, estimate$estimate, level),
    level = as.double(level), method = "level_set", radius = NULL
  ), class = "di_set")
}

#' Invert a simultaneous confidence band
#'
#' Returns grid points whose intervals contain the target level. This is the
#' alternative density level-set construction in Remark 2 and its inverse-
#' regression analogue.
#' @param band An object returned by a confidence-band function.
#' @param level Finite target level.
#' @return An object of class `di_set`.
#' @export
invert_confidence_band <- function(band, level) {
  if (!inherits(band, "di_band")) stop("band must be a di_band object", call. = FALSE)
  points <- band$points
  if (is.null(dim(points))) {
    points <- .as_vector(points, "band$points")
  } else {
    points <- .as_set_points(points, "band$points")
  }
  if (!is.numeric(level) || length(level) != 1L || !is.finite(level)) {
    stop("level must be finite", call. = FALSE)
  }
  structure(list(
    points = points, mask = band$lower <= level & level <= band$upper,
    roots = .level_geometry(points, band$estimate, level),
    level = as.double(level), method = "band_inversion", radius = NULL,
    confidence = band$confidence, n_boot = band$n_boot,
    bootstrap_statistics = band$bootstrap_statistics
  ), class = "di_set")
}

#' Hausdorff distance between finite point clouds
#' @param a First non-empty numeric vector or point matrix.
#' @param b Second non-empty numeric vector or point matrix.
#' @return A nonnegative scalar.
#' @export
hausdorff_distance <- function(a, b) {
  a <- .as_set_points(a, "a")
  b <- .as_set_points(b, "b")
  if (ncol(a) != ncol(b)) stop("a and b must have the same point dimension", call. = FALSE)
  distances <- matrix(0, nrow = nrow(a), ncol = nrow(b))
  for (coordinate in seq_len(ncol(a))) {
    differences <- outer(a[, coordinate], b[, coordinate], "-")
    distances <- distances + differences^2
  }
  distances <- sqrt(distances)
  max(max(apply(distances, 1L, min)), max(apply(distances, 2L, min)))
}

#' Estimate a density equality level set
#' @inheritParams debiased_kde
#' @param level Finite target density level.
#' @return An object of class `di_set`.
#' @export
density_level_set <- function(x, level, points = NULL, bandwidth = NULL,
                              tau = 1, grid_size = 200L) {
  level_set(debiased_kde(
    x, points, bandwidth = bandwidth, tau = tau, grid_size = grid_size
  ), level)
}

#' Confidence set for a density equality level set
#'
#' The Hausdorff method implements Section 3.1.1 of Cheng and Chen. The
#' inversion method implements the alternative construction in Remark 2.
#' @inheritParams kde_confidence_band
#' @param level Finite target density level.
#' @param method Either `"hausdorff"` or `"inversion"`.
#' @param max_attempts Maximum bootstrap samples attempted if an estimated
#'   level set is empty.
#' @return An object of class `di_set`.
#' @export
density_level_set_confidence <- function(
    x, level, points = NULL, bandwidth = NULL, tau = 1, confidence = 0.95,
    n_boot = 999L, method = "hausdorff", random_state = NULL,
    grid_size = 200L, max_attempts = NULL) {
  if (length(method) != 1L || is.na(method) ||
      !method %in% c("hausdorff", "inversion")) {
    stop("method must be 'hausdorff' or 'inversion'", call. = FALSE)
  }
  if (method == "inversion") {
    return(invert_confidence_band(kde_confidence_band(
      x, points, bandwidth = bandwidth, tau = tau, confidence = confidence,
      n_boot = n_boot, random_state = random_state, grid_size = grid_size
    ), level))
  }
  parameters <- .bootstrap_parameters(confidence, n_boot)
  confidence <- parameters$confidence
  n_boot <- parameters$n_boot
  samples <- .as_samples(x)
  if (!ncol(samples) %in% c(1L, 2L)) {
    stop("Hausdorff level-set confidence supports one or two dimensions",
         call. = FALSE)
  }
  evaluation_matrix <- .as_points(points, samples, grid_size)
  evaluation <- if (ncol(samples) == 1L) {
    evaluation_matrix[, 1L]
  } else {
    evaluation_matrix
  }
  if (is.null(bandwidth)) bandwidth <- density_bandwidth(samples)
  bandwidth <- .positive_scalar(bandwidth, "bandwidth")
  tau <- .positive_scalar(tau, "tau")
  if (!is.numeric(level) || length(level) != 1L || !is.finite(level)) {
    stop("level must be finite", call. = FALSE)
  }
  level <- as.double(level)
  kernel_matrix <- .kernel_matrix(samples, evaluation_matrix, bandwidth, tau)
  estimate <- .kde_values(kernel_matrix, bandwidth, ncol(samples))
  roots <- .level_geometry(evaluation, estimate, level)
  if (length(roots) == 0L) {
    stop("estimated level set is empty on the evaluation grid", call. = FALSE)
  }
  if (is.null(max_attempts)) max_attempts <- max(10L * n_boot, 100L)
  if (!is.numeric(max_attempts) || length(max_attempts) != 1L ||
      !is.finite(max_attempts) ||
      max_attempts < n_boot || max_attempts != as.integer(max_attempts)) {
    stop("max_attempts must be an integer at least n_boot", call. = FALSE)
  }
  statistics <- .with_seed(random_state, {
    result <- numeric(n_boot)
    accepted <- 0L
    attempts <- 0L
    while (accepted < n_boot && attempts < max_attempts) {
      attempts <- attempts + 1L
      indices <- sample.int(nrow(samples), nrow(samples), replace = TRUE)
      counts <- tabulate(indices, nbins = nrow(samples))
      bootstrap_estimate <- as.double(kernel_matrix %*% counts) /
        (nrow(samples) * bandwidth^ncol(samples))
      bootstrap_roots <- .level_geometry(evaluation, bootstrap_estimate, level)
      if (length(bootstrap_roots) > 0L) {
        accepted <- accepted + 1L
        result[accepted] <- hausdorff_distance(bootstrap_roots, roots)
      }
    }
    if (accepted < n_boot) {
      stop("only ", accepted, " non-empty bootstrap level sets after ",
           attempts, " attempts", call. = FALSE)
    }
    result
  })
  radius <- as.double(stats::quantile(
    statistics, confidence, names = FALSE, type = 7
  ))
  distance_to_roots <- .distance_to_geometry(evaluation, roots)
  structure(list(
    points = evaluation, mask = distance_to_roots <= radius, roots = roots,
    level = level, method = "hausdorff", radius = radius,
    confidence = confidence, n_boot = n_boot,
    bootstrap_statistics = statistics
  ), class = "di_set")
}

#' Estimate an inverse-regression equality set
#' @inheritParams debiased_local_linear
#' @param level Finite target response level.
#' @return An object of class `di_set`.
#' @export
inverse_regression <- function(x, y, level, points = NULL, bandwidth = NULL,
                               tau = 1, grid_size = 200L, n_folds = 5L,
                               random_state = 0L) {
  level_set(debiased_local_linear(
    x, y, points, bandwidth = bandwidth, tau = tau, grid_size = grid_size,
    n_folds = n_folds, random_state = random_state
  ), level)
}

#' Confidence set for an inverse-regression equality set
#' @inheritParams regression_confidence_band
#' @param level Finite target response level.
#' @param method One of `"hausdorff"`, `"inversion"`, or `"normal"`. The
#'   normal method requires exactly one estimated crossing.
#' @return An object of class `di_set`.
#' @export
inverse_regression_confidence <- function(
    x, y, level, points = NULL, bandwidth = NULL, tau = 1,
    confidence = 0.95, n_boot = 999L, method = "hausdorff",
    random_state = NULL, grid_size = 200L, n_folds = 5L,
    max_attempts = NULL) {
  if (length(method) != 1L || is.na(method) ||
      !method %in% c("hausdorff", "inversion", "normal")) {
    stop("method must be 'hausdorff', 'inversion', or 'normal'", call. = FALSE)
  }
  if (method == "inversion") {
    return(invert_confidence_band(regression_confidence_band(
      x, y, points, bandwidth = bandwidth, tau = tau,
      confidence = confidence, n_boot = n_boot,
      random_state = random_state, grid_size = grid_size,
      n_folds = n_folds, max_attempts = max_attempts
    ), level))
  }
  parameters <- .bootstrap_parameters(confidence, n_boot)
  confidence <- parameters$confidence
  n_boot <- parameters$n_boot
  x <- .as_vector(x, "x")
  y <- .as_vector(y, "y")
  if (length(x) != length(y) || length(x) < 4L) {
    stop("x and y must have equal lengths of at least 4", call. = FALSE)
  }
  base <- debiased_local_linear(
    x, y, points, bandwidth = bandwidth, tau = tau, grid_size = grid_size,
    n_folds = n_folds, random_state = random_state
  )
  if (!is.numeric(level) || length(level) != 1L || !is.finite(level)) {
    stop("level must be finite", call. = FALSE)
  }
  level <- as.double(level)
  roots <- .crossing_roots(base$points, base$estimate, level)
  if (length(roots) == 0L) {
    stop("estimated inverse-regression set is empty on the evaluation grid",
         call. = FALSE)
  }
  if (method == "normal" && length(roots) != 1L) {
    stop("method='normal' requires exactly one estimated crossing", call. = FALSE)
  }
  if (method == "normal" && n_boot < 2L) {
    stop("method='normal' requires at least two bootstrap replicates", call. = FALSE)
  }
  if (is.null(max_attempts)) max_attempts <- max(10L * n_boot, 100L)
  if (!is.numeric(max_attempts) || length(max_attempts) != 1L ||
      !is.finite(max_attempts) ||
      max_attempts < n_boot || max_attempts != as.integer(max_attempts)) {
    stop("max_attempts must be an integer at least n_boot", call. = FALSE)
  }
  statistics <- .with_seed(random_state, {
    result <- numeric(n_boot)
    accepted <- 0L
    attempts <- 0L
    while (accepted < n_boot && attempts < max_attempts) {
      attempts <- attempts + 1L
      indices <- sample.int(length(x), length(x), replace = TRUE)
      estimate <- .debiased_regression_values(
        x[indices], y[indices], base$points, base$bandwidth, base$tau
      )
      if (all(is.finite(estimate))) {
        bootstrap_roots <- .crossing_roots(base$points, estimate, level)
        valid_roots <- if (method == "normal") {
          length(bootstrap_roots) == 1L
        } else {
          length(bootstrap_roots) > 0L
        }
        if (valid_roots) {
          accepted <- accepted + 1L
          result[accepted] <- if (method == "normal") {
            bootstrap_roots[1L]
          } else {
            hausdorff_distance(bootstrap_roots, roots)
          }
        }
      }
    }
    if (accepted < n_boot) {
      stop("only ", accepted, " non-empty bootstrap inverse sets after ",
           attempts, " attempts", call. = FALSE)
    }
    result
  })
  radius <- if (method == "normal") {
    stats::qnorm(0.5 + confidence / 2) * stats::sd(statistics)
  } else {
    as.double(stats::quantile(statistics, confidence, names = FALSE, type = 7))
  }
  distance_to_roots <- apply(abs(outer(base$points, roots, "-")), 1L, min)
  structure(list(
    points = base$points, mask = distance_to_roots <= radius, roots = roots,
    level = level, method = method, radius = radius,
    confidence = confidence, n_boot = n_boot,
    bootstrap_statistics = statistics
  ), class = "di_set")
}
